### PR TITLE
Make some binary reading/writing utilities public

### PR DESCRIPTION
**Pull request by sbc100**
This is mostly a refactor to make some of the internal
binary reading/writing utilities public and available
to the wasm linker prototype that I'm working on.

### CODE REVIEW

File: src/binary-reader-objdump.c [hunk 1]
```diff
@@ -608,16 +595,16 @@ static WasmBinaryReader s_binary_reader = {
     .end_module = end_module,
     .on_error = on_error,
 
+    .begin_section = begin_section,
```
Comment (binji) [inline src/binary-reader-objdump.c]: seems weird to have a begin_section without an end_section

File: src/binary-writer.c [hunk 2]
```diff
@@ -90,31 +91,46 @@ static uint32_t write_u32_leb128_at(WasmStream* stream,
...
-static void write_u32_leb128(WasmStream* stream,
-                             uint32_t value,
-                             const char* desc) {
-  uint32_t length = write_u32_leb128_at(stream, stream->offset, value, desc);
+uint32_t wasm_write_fixed_u32_leb128_at(WasmStream* stream,
+                                        uint32_t offset,
+                                        uint32_t value,
+                                        const char* desc) {
+  uint8_t data[MAX_U32_LEB128_BYTES];
+  uint32_t rtn = wasm_write_fixed_u32_leb128_raw(data, data+MAX_U32_LEB128_BYTES, value);
```
Comment (binji) [inline src/binary-writer.c]: nit: better name than `rtn`? How about `length`?

File: src/binary-writer.c [hunk 3]
```diff
@@ -180,38 +196,38 @@ static void write_fixup_u32_leb128_size(Context* ctx,
...
 }
 
-static void write_str(WasmStream* stream,
-                      const char* s,
-                      size_t length,
-                      WasmPrintChars print_chars,
-                      const char* desc) {
-  write_u32_leb128(stream, length, "string length");
+void wasm_write_str(WasmStream* stream,
+                           const char* s,
```
Comment (binji) [inline src/binary-writer.c]: nit: format

File: src/binary-writer.c [hunk 4]
```diff
@@ -272,176 +288,177 @@ static void write_expr(Context* ctx,
...
     }
     case WASM_EXPR_TYPE_STORE: {
-      write_opcode(&ctx->stream, expr->store.opcode);
+      wasm_write_opcode(&ctx->stream, expr->store.opcode);
       uint32_t align =
           wasm_get_opcode_alignment(expr->store.opcode, expr->store.align);
       wasm_write_u8(&ctx->stream, log2_u32(align), "alignment");
-      write_u32_leb128(&ctx->stream, (uint32_t)expr->store.offset,
+      wasm_write_u32_leb128(&ctx->stream, (uint32_t)expr->store.offset,
                        "store offset");
```
Comment (binji) [inline src/binary-writer.c]: nit: format
other places in this file too

File: src/binary-writer.h [hunk 5]
```diff
@@ -18,6 +18,7 @@
 #define WASM_BINARY_WRITER_H_
 
 #include "common.h"
+#include "stream.h"
```
Comment (binji) [inline src/binary-writer.h]: Can you forward declare WasmStream instead?

Comment (sbc100) [inline src/binary-writer.h]: Interestingly it was an enum I needed.. and I didn't realize I could forward declare them.

Done.

File: src/binary-reader-objdump.c (hunk 1 above)
Comment (sbc100) [inline src/binary-reader-objdump.c]: Should I add that then, even though there are no consumers for it yet?

Comment (binji) [inline src/binary-reader-objdump.c]: I'd prefer it, but maybe that's just my OCD... :)

**Pull request merged**